New Post: Chromium Lands Flexbox Gap (#2759)
GoogleChrome/web.dev: src/site/content/en/blog/flexbox-gap/index.md

---

> title: Chromium Lands Flexbox `gap`

Reviewer: ```suggestion
title: Chromium lands Flexbox `gap`
```
We use sentence case for all titles

> subhead: Chromium's CSS Flexbox and Multi-Column layout engines now support gap. `grid-gap` is old news. 

Reviewer: ```suggestion
subhead: Chromium's CSS Flexbox and Multi-Column layout engines now support `gap`. `grid-gap` is old news. 
```

authors:
  - adamargyle
  # - dgrogan

> description: Chromium's CSS Flexbox and Multi-Column layout engines now support gap. `grid-gap` is old news. 

Reviewer: ```suggestion
description: Chromium's CSS Flexbox and Multi-Column layout engines now support `gap`. `grid-gap` is old news. 
```

date: 2020-04-30

> updated: 2020-04-30

Reviewer: ```suggestion
```
I think we can omit this until we actually update the guide

tags:
  - layout
  - flexbox
  - css
  - gap

> # hero: hero.jpg

Reviewer: We use unsplash.com for most of our hero images

Author: found one 👍

alt: TODO
---

CSS Grid brought a delightful spacing feature called [`grid-gap`](https://developer.mozilla.org/en-US/docs/Web/CSS/gap), which quickly became a popular way to put space into layouts. It was easy to fall in love with because it took upon itself so much responsibility with such elegance. Starting in Chromium 85 `grid-gap` is now just [`gap`](https://drafts.csswg.org/css-align-3/#propdef-gap), and you can use it with Flexbox. 💪

## Browser compatibility

At the time of writing `gap` is supported in desktop Firefox (version 63), Firefox for Android (version 63), and Chromium-based browsers (version 85). See [Browser compatibility](https://developer.mozilla.org/docs/Web/CSS/gap#Browser_compatibility) for updates.

```css/3-4/1-2
.grid {
  display: grid;
  grid-gap: 10px;
  display: flex;
  gap: 10px;
}

> ```

Reviewer: This code block seems out-of-place. I think we want it before the `Browser compatibility` section?

![image](https://user-images.githubusercontent.com/4713486/80765414-d0089500-8af7-11ea-9409-5eef8b883ed3.png)

Author: done 👍

## CSS Gap

{% Aside 'key-term' %}
`gap` is the spacing *between children*. You may have heard of this type of spacing being called "gutters" or "alleys". It's space only where the children box edges touch.
{% endAside %}

`gap` is [flow relative](https://www.w3.org/TR/css-logical-1/#intro), meaning it changes dynamically based on the direction of content flow. For example, `gap` will automatically adjust for the different`writing-mode` or `direction` values that you set for your international users. This significantly eases the burden of spacing challenges for the component and CSS author. **Less code scaling further.**

<figure class="w-figure">
  <video controls autoplay loop muted class="w-screenshot">

>     <!-- <source src="https://storage.googleapis.com/web-dev-assets/portals_vp9.webm" type="video/webm; codecs=vp8"> -->

Reviewer: Do you need me to upload this? Link me to the webm file on Drive if so, plz

Author: i can upload, i need to convert

Author: unless you wanna download the mp4, convert, and upload for me? i'm equipped for the task, but happy to get help?

Reviewer: I tried to halp but `ffmpeg` is taking forever on my machine

Set both rows and columns at once
{% endCompareCaption %}

{% endCompare %}

{% Compare 'better', 'Expanded' %}
```css
.grid {
  display: grid;
  row-gap: 10px;
  column-gap: 10px;
}
```

{% CompareCaption %}
Set them separately
{% endCompareCaption %}

{% endCompare %}
</div>

<br>

Gap can be passed 2 lengths, which will be used for row and column.

<div class="w-columns">
{% Compare 'better', 'Shorthand' %}
```css
.grid {
  display: grid;
  gap: 10px 5%;
}
```

{% CompareCaption %}
Set both rows and columns **separately** at once
{% endCompareCaption %}

{% endCompare %}

{% Compare 'better', 'Expanded' %}
```css
.grid {
  display: grid;
  row-gap: 10px;
  column-gap: 5%;
}
```

{% endCompare %}
</div>

## Flexbox `gap`

All of that `gap` goodness we just reviewed is available with Flexbox containers now! Before `gap` was in Flexbox, strategies involved negative margins, complex selectors, `:last` or `:first` type pseudo-class selectors, or other means to manage the space of a dynamically layed-out and wrapping set of children.

### Previous Attempts

Example #1 - pseudo-class selectors
```css
.layout > :not(:first-child) {
  margin-block-end: 10px;
  margin-inline-end: 5%;
}

> ```

Reviewer: How about styling this with the `{% Compare 'worse' %}` component? You can give it a custom label if you want to use a gentler word. https://web.dev/handbook/web-dev-components/#compare-with-custom-labels

Author: done 👍

Example #2 - lobotomized owl selector
```css
.layout > * + * {
  margin-block-end: 10px; 
  margin-inline-end: 5%; 
}

> ```

Reviewer: Same comment as example 1

The above are not a full replacement for `gap` though, and often need `@media` query adjustments to account for wrapping scenarios or writing modes or direction. Adding 1 or 2 media queries doesn't seem so bad, but they can add up and lead to complicated layout logic.

What the above author really intended was to have none of the child items touch. 

> s

Reviewer: ```suggestion

```

## Chromium DevTools Updates
With these updates come changes to Chromium DevTools, notice how the **Styles** pane handles `grid-gap` and `gap` now 👍

<figure class="w-figure">
  <img src="./devtools-gap.png" alt="An office with two people working at a table." style="max-width: 400px; border-radius: 10px;">
  <figcaption class="w-figcaption">Devtools shows the both grid-gap and gap, with gap shown used below grid-gap as to let the cascade use the latest syntax.</figcaption>
</figure>

DevTools supports both `grid-gap` and `gap`, this is because `gap` is essentially an alias to the previous syntaxes. 

## New layout potential

With `flex gap`, we unlock more than convenience, we unlock powerful, perfectly spaced, intrinsic layouts. In the video and following code sample below, `grid` cannot achieve the layout that `flex` can. `grid` must have equal rows and columns, even if they're intrinsically assigned. 

<figure class="w-figure">
  <video controls autoplay loop muted class="w-screenshot">

>     <!-- <source src="https://storage.googleapis.com/web-dev-assets/portals_vp9.webm" type="video/webm; codecs=vp8"> -->

Reviewer: Link me to the Drive file if you need me to upload

---
title: Chromium lands Flexbox `gap`
subhead: Chromium's CSS Flexbox and Multi-Column layout engines now support `gap`. `grid-gap` is old news. 
authors:
  - adamargyle
  - dgrogan
description: Chromium's CSS Flexbox and Multi-Column layout engines now support `gap`. `grid-gap` is old news. 
date: 2020-04-30

> tags:
>   - layout

Reviewer: ```suggestion
tags:
  - blog
  - layout
```

---
title: Chromium lands Flexbox `gap`
subhead: Chromium's CSS Flexbox and Multi-Column layout engines now support `gap`. `grid-gap` is old news. 
authors:
  - adamargyle
  - dgrogan
description: Chromium's CSS Flexbox and Multi-Column layout engines now support `gap`. `grid-gap` is old news. 

> date: 2020-04-30

Reviewer: ```suggestion
date: 2020-05-06
```

All of that `gap` goodness we just reviewed is available with Flexbox containers now! Before `gap` was in Flexbox, strategies involved negative margins, complex selectors, `:last` or `:first` type pseudo-class selectors, or other means to manage the space of a dynamically layed-out and wrapping set of children.

### Previous Attempts
The following are patterns that folks have used to get gap-like spacing.

<div class="w-columns">
{% Compare 'worse', 'pseudo-class selectors' %}
```css
.layout > :not(:last-child) {
  margin-bottom: 10px;
  margin-right: 10px;
}
```

{% endCompare %}

{% Compare 'worse', 'lobotomized owl' %}
```css
.layout > * + * {
  margin-bottom: 10px; 
  margin-right: 10px; 
}
```

{% CompareCaption %}
[Source](https://alistapart.com/article/axiomatic-css-and-lobotomized-owls/)
{% endCompareCaption %}

{% endCompare %}
</div>

> The above are not a full replacement for `gap` though, and often need `@media` query adjustments to account for wrapping scenarios, writing modes or direction. Adding 1 or 2 media queries doesn't seem so bad, but they can add up and lead to complicated layout logic.

Reviewer 2: ```suggestion
The above are not a full replacement for `gap` though, and often need `@media` query adjustments to account for wrapping scenarios, writing modes or direction. Adding one or two media queries doesn't seem so bad, but they can add up and lead to complicated layout logic.
```
